src/heaps/fibonacci_heap.jl

```diff
@@ -0,0 +1,367 @@
+    if node == root.rootList
+        root.rootList = node.right
+    end
+    node.left.right = node.right
+    node.right.left = node.left
+end
+
+# remove a node from the doubly linked child list
+function _remove_from_child_list(root::MutableFibonacciHeap{T}, parent::MutableFibonacciHeapNode{T}, node::MutableFibonacciHeapNode{T}) where T
+    if parent.child == parent.child.right
+        parent.child = nothing
+    elseif parent.child == node
+        parent.child = node.right
+        node.right.parent = parent
+    end
+    node.left.right = node.right
+    node.right.left = node.left
+end
+
+# merge a node with the doubly linked child list of a root node
+function _merge_with_child_list(root::MutableFibonacciHeap{T}, parent::MutableFibonacciHeapNode{T}, node::MutableFibonacciHeapNode{T}) where T
+    if parent.child == nothing
+        parent.child = node
+    else
+        node.right = parent.child.right
+        node.left = parent.child
+        parent.child.right.left = node
+        parent.child.right = node
+    end
+end
+
+# extract (delete) the min node from the heap in O(log n) time
+function extractMin(root::MutableFibonacciHeap)
+    minimumNode = root.minNode
+    if minimumNode !== nothing
+        if minimumNode.child !== nothing
+            # attach child nodes to root list
+            allChild = aslist(minimumNode.child)
+            for child in allChild
+                _merge_with_root_list(root, child)
+                child.parent = nothing
+            end
+        end
+        
+        _remove_from_root_list(root, minimumNode)
+        # set new min node in heap
+        if minimumNode == minimumNode.right
+            root.minNode = root.rootList = nothing
+        else
+            root.minNode = minimumNode.right
+            consolidate(root)
+        end
+        root.nodeCount -= 1
+    end
+    return minimumNode.value
+end
+
+# actual linking of one node to another in the root list
+# while also updating the child linked list
+function heaplink(root::MutableFibonacciHeap{T}, y::MutableFibonacciHeapNode{T}, x::MutableFibonacciHeapNode{T}) where T
+    _remove_from_root_list(root, y)
+    y.left = y.right = y
+    _merge_with_child_list(root, x, y)
+    x.degree += 1
+    y.parent = x
+    y.mark = false
+end
+
+# combine root nodes of equal degree to consolidate the heap
+# by creating a list of unordered binomial trees
+function consolidate(root::MutableFibonacciHeap{T}) where T
+    nodes = aslist(root.rootList)
+    len = length(nodes)
+    A = fill!(Vector{Union{MutableFibonacciHeapNode, Nothing}}(undef, Int(round(log(root.nodeCount)) * 2)), nothing)
+    for x in nodes
+        d = x.degree + 1
+        while A[d] !== nothing
+            y = A[d]
+            if x.value > y.value
+                x, y = y, x
+            end
+            heaplink(root, y, x)
+            A[d] = nothing
+            d += 1
+        end
+        A[d] = x
+    end
```

Comment:
It might be worth adding `@inbounds` to this loops if it's a hot spot. Not sure though!